python/pyarrow/tests/test_dataset.py

```diff
@@ -206,6 +206,63 @@ def dataset(mockfs):
     return factory.finish()
 
 
+@pytest.fixture(params=[
+    (True, True),
+    (True, False),
+    (False, True),
+    (False, False)
+], ids=['threaded-async', 'threaded-sync', 'serial-async', 'serial-sync'])
+def dataset_reader(request):
+    '''
+    Fixture which allows dataset scanning operations to be
+    run with/without threads and with/without async
+    '''
+    use_threads, use_async = request.param
+
+    class reader:
+
+        def __init__(self):
+            self.use_threads = use_threads
+            self.use_async = use_async
+
+        def _patch_kwargs(self, kwargs):
+            if 'use_threads' in kwargs:
+                raise Exception(
+                    ('Invalid use of dataset_reader, do not specify'
+                     ' use_threads'))
+            if 'use_async' in kwargs:
+                raise Exception(
+                    'Invalid use of dataset_reader, do not specify use_async')
+            kwargs['use_threads'] = use_threads
+            kwargs['use_async'] = use_async
+
+        def to_table(self, dataset, **kwargs):
+            self._patch_kwargs(kwargs)
+            return dataset.to_table(**kwargs)
+
+        def to_batches(self, dataset, **kwargs):
+            self._patch_kwargs(kwargs)
+            return dataset.to_batches(**kwargs)
+
+        def scanner(self, dataset, **kwargs):
+            self._patch_kwargs(kwargs)
+            return dataset.scanner(**kwargs)
+
+        def head(self, dataset, num_rows, **kwargs):
```

Comment:
Note you could declare all these as `(self, dataset, *args, **kwargs)` and avoid having to replicate their individual signatures.

Reply:
Thanks for converting this file, by the way, this looks like it was rather tedious.

Reply:
I considered this but felt this was more self documenting.  I don't have a strong opinion though, so if you want I'm happy to switch it.

I also realized I could just monkeypatch the dataset/fragment but most tests create their dataset/fragment outside of pytest so that would end up being just as much of a change.

Reply:
I think the way it is is fine too.